pkg/controller/pxc/upgrade.go

```diff
@@ -115,7 +126,156 @@ func (r *ReconcilePerconaXtraDBCluster) updatePod(sfs api.StatefulApp, podSpec *
 	currentSet.Spec.Template.Spec.Affinity = pxc.PodAffinity(podSpec.Affinity, sfs)
 	currentSet.Spec.Template.Spec.Volumes = sfsVolume.Volumes
 
-	return r.client.Update(context.TODO(), currentSet)
+	err = r.client.Update(context.TODO(), currentSet)
+	if err != nil {
+		return fmt.Errorf("update error: %v", err)
+	}
+
+	if !cr.Spec.SmartUpdateEnabled() {
+		return nil
+	}
+
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: currentSet.Name, Namespace: currentSet.Namespace}, currentSet)
+	if err != nil {
+		return fmt.Errorf("failed to get sate: %v", err)
+	}
+
+	if currentSet.Generation == startGeneration {
+		return nil
+	}
+
+	log.Info("statefullSet was changed, run smart update")
+
+	return r.smartUpdate(sfs, cr)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) smartUpdate(sfs api.StatefulApp, cr *api.PerconaXtraDBCluster) error {
+	if !isPXC(sfs) {
+		return nil
+	}
+
+	primary, err := r.getPrimaryPod(cr)
+	if err != nil {
+		return fmt.Errorf("get primary pod: %v", err)
+	}
+
+	log.Info(fmt.Sprintf("primary pod is %s", primary))
+
+	list := corev1.PodList{}
+	err = r.client.List(context.TODO(),
+		&client.ListOptions{
+			Namespace:     sfs.StatefulSet().Namespace,
+			LabelSelector: labels.SelectorFromSet(sfs.Labels()),
+		},
+		&list,
+	)
+	if err != nil {
+		return fmt.Errorf("get pod list: %v", err)
+	}
+
+	var primaryPod *corev1.Pod = nil
+	for _, pod := range list.Items {
+		pod := pod
+		if strings.HasPrefix(primary, pod.Name) {
+			primaryPod = &pod
+		} else {
+			log.Info(fmt.Sprintf("delete secondary pod %s", pod.Name))
+			err := r.client.Delete(context.TODO(), &pod)
+			if err != nil {
+				return fmt.Errorf("delete pod: %v", err)
+			}
+
+			err = r.waitPodRestart(&pod)
+			if err != nil {
+				return fmt.Errorf("wait pod: %v", err)
+			}
+		}
+	}
+
+	log.Info(fmt.Sprintf("delete primary pod %s", primaryPod.Name))
+	err = r.client.Delete(context.TODO(), primaryPod)
+	if err != nil {
+		return fmt.Errorf("delete primary pod: %v", err)
+	}
+
+	err = r.waitPodRestart(primaryPod)
+	if err != nil {
+		return fmt.Errorf("wait pod: %v", err)
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) getPrimaryPod(cr *api.PerconaXtraDBCluster) (string, error) {
+	secretObj := corev1.Secret{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      cr.Spec.SecretsName,
+		},
+		&secretObj,
+	)
+	if err != nil {
+		return "", err
+	}
+
+	user := "proxyadmin"
+	pass := string(secretObj.Data[user])
+	host := fmt.Sprintf("%s-proxysql-unready", cr.ObjectMeta.Name)
```

Comment:
multiple namespaces case
```suggestion
	host := fmt.Sprintf("%s-proxysql-unready.%s", cr.ObjectMeta.Name, cr.Namespace)
```